### ISSUE TITLE
Drag from iPhoto for image not copied to library fails.

### ISSUE THREAD

**Issue by danwood** [karelia/iMedia#22]
If you turn off the advanced preference in iPhoto to "copy items to the iPhoto library", and put an image into your iphoto library....

Then try dragging that image, from iPhoto library into some folder (other than the original file path where the file actually resides)

then somehow the copy does not happen.

I'm stumped!  The "ImagePath" in the iphoto library file is just the absolute path of the image, and that's valid.  Our code stores that into the "location" property of the IMBNodeObject.

I'm not having any luck finding the code where this path is then resolved into the source path for actually copying the file.  Help!

**Comment by mikeabdullah** [comment]
I've been hunting this one down. iPhoto reports that all photos are located inside of its Library package, even if you've chosen not to copy the image in.

Why? Because what iPhoto does in those circumstances is create an _alias_ to the original image, inside of its package. I guess this gives them the advantage of using a fairly standard component that can track the image if it gets moved.

How should we tackle this? It seems iMedia needs to resolve any aliases before handing image paths off to the pasteboard. How?
- OS X 10.6 has Cocoa API for resolving aliases, we could choose to support aliases only on Snowy
- Earlier OS's have only Carbon-level support, we could write the extra code to use this
- Third party wrappers around the Alias Manager exist, we could include one of them in the library

There's no indication from iPhoto whether a photo is an alias or a real image. We have to test every file to see if it is an alias. Doing this for a whole library at once is clearly a no-no! Perhaps each `IMBObject` could take on responsibility for performing a one-off, on-demand test?

**Comment by gloubibou** [comment]
I did some work resolving aliased images in IMBiPhotoObjectPromise

**Comment by peterb180369** [comment]
Yes that sounds like the correct (lazy) approach to me - only do the resolving at the last moment (in IMBiPhotoObjectPromise), when photos were actually dragged from the browser and dropped at a receiving destination. That way we only do the required amount of work.

**Comment by mikeabdullah** [comment]
Ah, so I see the alias handling in `IMBiPhotoObjectPromise`, thanks! However it only kicks in for photos in `Masters` as I understand it. In my case pretty certain the photo is in `Originals` so doesn't have this logic applied. I'll experiment with expanding that.

**Comment by mikeabdullah** [comment]
Have a look at the bugfix/22 branch which resolves this bug for us. Does this seem a rather brute force solution to you guys? I don't really understand what the `Masters` bit of the code is about; we just need the object's location to be updated so the promise references it correctly.

Is there any potential problem with mutating an object's location? If so, why isn't this a readonly property?

**Comment by gloubibou** [comment]
Hi!

Here is what the Masters/ bit of code is about.
1. Set iPhoto not to copy images into the library
2. Load images from an external drive
3. Quit iPhoto

AlbumData.xml now references the images at their original location
1. Take the external drive offline
2. Launch iPhoto 
3. Quit iPhoto

AlbumData.xml now references aliases to the offline images. These aliases are located at Masters/ within the library.
This makes the images appear as if they were hosted by the library when they are not.

**Comment by mikeabdullah** [comment]
I see. So would my more generic code handle this case too? I notice you're adjusting the metadata dictionary, not the object's location, so suspect not.

**Comment by gloubibou** [comment]
The paths my code tries to fix start as relative paths Masters/SomeAlias.jpg. I first need to convert those to absolute paths before resolving aliases.

I update the metadata because that is what HoudahGeo looks for when receiving the dragged file. It needs the preview and master paths.
